FSharpLu.Json/Compact.fs

```diff
@@ -25,67 +63,67 @@ open ConverterHelpers
 type CompactUnionJsonConverter() =
     inherit Newtonsoft.Json.JsonConverter()
 
-    let SomeFieldIdentifier = "Some"
-
-    /// Determine if a given type has a field named 'Some' which would cause
-    /// ambiguity if nested under an option type without being boxed
-    let hasFieldNamedSome (t:System.Type) =
-        isOptionType t // the option type itself has a 'Some' field
-        || (FSharpType.IsRecord t && FSharpType.GetRecordFields t |> Seq.exists (fun r -> stringEq r.Name SomeFieldIdentifier))
-        || (FSharpType.IsUnion t && FSharpType.GetUnionCases t |> Seq.exists (fun r -> stringEq r.Name SomeFieldIdentifier))
+    let canConvertMemorisation = Dictionary<_, _>()
+    let optionTypeMemorisation = Dictionary<_, _>()
 
     override __.CanConvert(objectType:System.Type) =
-        // Include F# discriminated unions
-        FSharpType.IsUnion objectType
-        // and exclude the standard FSharp lists (which are implemented as discriminated unions)
-        && not (objectType.GetTypeInfo().IsGenericType && objectType.GetGenericTypeDefinition() = typedefof<_ list>)
+        memorise 
+            (fun objectType ->        
+                // Include F# discriminated unions
+                let result = 
+                    FSharpType.IsUnion objectType
+                    // and exclude the standard FSharp lists (which are implemented as discriminated unions)
+                    && not (objectType.GetTypeInfo().IsGenericType && objectType.GetGenericTypeDefinition() = typedefof<_ list>)
+                result)
+            canConvertMemorisation
+            objectType
+
+    static member WriteOptionType<'t> (writer: JsonWriter) (serializer:JsonSerializer) (param: obj) = 
+        let x = param :?> 't option
 
-    override __.WriteJson(writer:JsonWriter, value:obj, serializer:JsonSerializer) =
         let convertName =
             match serializer.ContractResolver with
             | :? CamelCasePropertyNamesContractResolver -> toCamel
             | _ -> id
+        
+        match x with
+        | None -> () // Just serialise as Null
+        // Note this is different than None. This case is to deal with C# created objects
+        | Some(v) when obj.ReferenceEquals(null, v) ->
```

Comment:
Will change that. Initially I was passing in the "param" arg as a ('t option) and because I would then be matching for Some(Some(v)) after this match with that the "isNull" would not type check as obviously options can't be null in pure F# code. With the change to (param: obj) can now use the isNull check